toolchain/check/eval.cpp

```diff
@@ -943,23 +933,161 @@ static auto PerformBuiltinBinaryIntOp(Context& context, SemIRLoc loc,
+                          TypedInt, BoolAsSelect, TypedInt);
+        context.emitter().Emit(
+            loc, CompileTimeShiftNegative,
+            {.type = lhs.type_id, .value = lhs_val},
+            builtin_kind == SemIR::BuiltinFunctionKind::IntLeftShift,
+            {.type = rhs.type_id, .value = rhs_orig_val});
+        // TODO: Is it useful to recover by returning 0 or -1?
+        return SemIR::ErrorInst::SingletonConstantId;
+      }
+
+      llvm::APInt result_val;
+      if (builtin_kind == SemIR::BuiltinFunctionKind::IntLeftShift) {
+        if (!lhs_bit_width_id.is_valid() && !lhs_val.isZero()) {
+          // Ensure we don't generate a ridiculously large integer through a bit
+          // shift.
+          auto width = rhs_orig_val.trySExtValue();
+          if (!width ||
+              *width > IntStore::MaxIntWidth - lhs_val.getSignificantBits()) {
+            CARBON_DIAGNOSTIC(CompileTimeUnsizedShiftOutOfRange, Error,
+                              "shift distance of {0} would result in an "
+                              "integer whose width is greater than the "
+                              "maximum supported width of {1}",
+                              TypedInt, int);
+            context.emitter().Emit(loc, CompileTimeUnsizedShiftOutOfRange,
+                                   {.type = rhs.type_id, .value = rhs_orig_val},
+                                   IntStore::MaxIntWidth);
+            return SemIR::ErrorInst::SingletonConstantId;
+          }
+          lhs_val = lhs_val.sext(IntStore::CanonicalBitWidth(
+              lhs_val.getSignificantBits() + *width));
+        }
+
+        result_val =
+            lhs_val.shl(rhs_orig_val.getLimitedValue(lhs_val.getBitWidth()));
+      } else if (lhs_is_signed) {
+        result_val =
+            lhs_val.ashr(rhs_orig_val.getLimitedValue(lhs_val.getBitWidth()));
+      } else {
+        CARBON_CHECK(lhs_bit_width_id.is_valid(),
+                     "Logical shift on unsized int");
+        result_val =
+            lhs_val.lshr(rhs_orig_val.getLimitedValue(lhs_val.getBitWidth()));
+      }
+      return MakeIntResult(context, lhs.type_id, lhs_is_signed,
+                           std::move(result_val));
+    }
+
+    default:
+      // Break to do additional setup for other builtin kinds.
+      break;
+  }
+
+  // Other operations are already checked to be homogeneous, so we can extend
+  // the RHS with the LHS bit width.
+  CARBON_CHECK(rhs.type_id == lhs.type_id, "Heterogeneous builtin integer op!");
+  auto [lhs_val, rhs_val] =
+      GetIntsAtSuitableWidth(context, lhs_bit_width_id, lhs.int_id, rhs.int_id);
+
+  BinaryIntOpResult result =
+      ComputeBinaryIntOpResult(builtin_kind, lhs_val, rhs_val);
+
+  if (result.overflow && !lhs_bit_width_id.is_valid()) {
+    // Retry with a larger bit width. Most operations can only overflow by one
+    // bit, but signed n-bit multiplication can overflow to 2n-1 bits.
+    int new_width =
```

Comment:
I originally planned to do that, but @chandlerc suggested it'd be better to do it this way -- IIUC the rationale is that we'll almost never need to go to a wider size than the inputs (because they've already been rounded up to a multiple of 64 bits by the `IntStore`), so it's better to speculatively assume that the result will fit than to spend time computing a width -- especially because any wider upper bound will require a heap allocation (`APInt` heap allocates integers wider than 64 bits.)